### PR TITLE
Add `to_date` function

### LINKED ISSUE #8987: Add a to_date function
### Is your feature request related to a problem or challenge?

In a similar vein to #8886 I believe it would be useful to have a to_date function that mirrors the functionality found in [postgresql](https://www.postgresql.org/docs/current/functions-formatting.html) and [spark](https://spark.apache.org/docs/latest/api/sql/#to_date). The format options however would follow the same pattern as #8886 and use the chrono library's formats (at least for the initial implementation)

Passing a single argument to the function would behave in a similar manner to the to_timestamp function where it was attempt to infer the date from either a number type (inferred as seconds/millis/etc since epoch), a timestamp or a utf8/largeutf8 string.  arrow-cast has a parse_date function invoked from the cast from string -> date32 that seems to handle a few common formats that can be documented in the user guide and examples.

The return type of this new function would be Date32 for arrow, Date for SQL.

### Describe the solution you'd like

A to_date(expr[,.format_n]) function that can be invoked from a dataframe or via sql in the same manner as to_timestamp.

### Describe alternatives you've considered

The basic implementation can likely be done using arrow_cast functionality currently however having a full function would allow for additional formatting options.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Tangruilin**
## Which issue does this PR close?

Closes #8987 .

## What changes are included in this PR?
Add to_date function, and change the protobuf file

## Are these changes tested?
Not now, but it should be added in this PR.

## Are there any user-facing changes?
Add a function named `to_date`, users can use `select to_date(1243124);`  to try it.

### CODE REVIEW

File: datafusion/proto/src/logical_plan/from_proto.rs
Comment (Tangruilin) [inline datafusion/proto/src/logical_plan/from_proto.rs:1817]: Here, I am not sure how to do with it, I need some suggestion

**Comment by Tangruilin** [comment]
@Omega359 
Hey! you can look it if you have time, I need you suggestion about the protobuf file and if there are something that i do not think about.

If it looks ok, I will write test code these days.

Thanks very much

**Comment by Omega359** [comment]
I haven't had time to look it over yet but first thing I did see was the use of Date64 - I can't see any good reason to use that vs Date32

**Comment by Tangruilin** [comment]
> I haven't had time to look it over yet but first thing I did see was the use of Date64 - I can't see any good reason to use that vs Date32

Date64 contains more number in my opinion. Maybe Date32 is better?

File: datafusion/proto/src/logical_plan/from_proto.rs
Comment (Weijun-H) [inline datafusion/proto/src/logical_plan/from_proto.rs:1817]: It should be
```suggestion
                ScalarFunction::ToDate => Ok(to_date(parse_expr(&args[0], registry)?)),
```

File: datafusion/physical-expr/src/datetime_expressions.rs
Comment (Weijun-H) [inline datafusion/physical-expr/src/datetime_expressions.rs:393]:
```suggestion

```

**Comment by Omega359** [comment]
> Date64 contains more number in my opinion. Maybe Date32 is better?

```
    /// A signed 32-bit date representing the elapsed time since UNIX epoch (1970-01-01)
    /// in days (32 bits).
    Date32,
    /// A signed 64-bit date representing the elapsed time since UNIX epoch (1970-01-01)
    /// in milliseconds (64 bits). Values are evenly divisible by 86400000.
    Date64,
```
Date32 is the way.

**Comment by Tangruilin** [comment]
> > Date64 contains more number in my opinion. Maybe Date32 is better?
> 
> ```
>     /// A signed 32-bit date representing the elapsed time since UNIX epoch (1970-01-01)
>     /// in days (32 bits).
>     Date32,
>     /// A signed 64-bit date representing the elapsed time since UNIX epoch (1970-01-01)
>     /// in milliseconds (64 bits). Values are evenly divisible by 86400000.
>     Date64,
> ```
> 
> Date32 is the way.

Okey! I will change it to Date32
I may finish it these two days

**Comment by Tangruilin** [comment]
@alamb @Omega359 This PR is ready to be reviewed

File: datafusion/physical-expr/src/datetime_expressions.rs
Comment (Weijun-H) [inline datafusion/physical-expr/src/datetime_expressions.rs:1352]:
```suggestion
```

File: datafusion/physical-expr/src/datetime_expressions.rs
Comment (Weijun-H) [inline datafusion/physical-expr/src/datetime_expressions.rs:1356]:
```suggestion
        return exec_err!(
```

File: datafusion/physical-expr/src/datetime_expressions.rs
Comment (Weijun-H) [inline datafusion/physical-expr/src/datetime_expressions.rs:1378]:
```suggestion
        DataType::Int32
        | DataType::Int64
        | DataType::Null
        | DataType::Float64
        | DataType::Date32
        | DataType::Date64 => cast_column(&args[0], &DataType::Date32, None),
```

File: datafusion/sqllogictest/test_files/dates.slt
Comment (Weijun-H) [inline datafusion/sqllogictest/test_files/dates.slt:131]: add a test for wrong input and return NULL

File: datafusion/physical-expr/src/datetime_expressions.rs
Comment (Weijun-H) [inline datafusion/physical-expr/src/datetime_expressions.rs:419]:
```suggestion
```

File: datafusion/physical-expr/src/datetime_expressions.rs
Comment (Weijun-H) [inline datafusion/physical-expr/src/datetime_expressions.rs:424]:
```suggestion
                            internal_datafusion_err!("Unable to cast to Date32")
```

File: datafusion/physical-expr/src/datetime_expressions.rs
Comment (Weijun-H) [inline datafusion/physical-expr/src/datetime_expressions.rs:408]:
```suggestion
                            internal_datafusion_err!("Unable to cast to Date32")
```

**Review by Weijun-H — COMMENTED** [review]
Thanks @Tangruilin I left some minor suggestions.

File: datafusion/sqllogictest/test_files/dates.slt
Comment (Tangruilin) [reply datafusion/sqllogictest/test_files/dates.slt:131]: Now the wrong input will return error but not NULL like `to_timestamp`

File: datafusion/sqllogictest/test_files/dates.slt
Comment (Tangruilin) [inline datafusion/sqllogictest/test_files/dates.slt:133]: refer to this, this is a example for wrong input

Comment (Tangruilin) [reply datafusion/sqllogictest/test_files/dates.slt:133]: @Omega359

File: datafusion/proto/src/logical_plan/from_proto.rs
Comment (Tangruilin) [reply datafusion/proto/src/logical_plan/from_proto.rs]: Done

File: datafusion/physical-expr/src/datetime_expressions.rs
Comment (Tangruilin) [reply datafusion/physical-expr/src/datetime_expressions.rs]: Done

File: datafusion/physical-expr/src/datetime_expressions.rs
Comment (Tangruilin) [reply datafusion/physical-expr/src/datetime_expressions.rs]: Done

File: datafusion/physical-expr/src/datetime_expressions.rs
Comment (Tangruilin) [reply datafusion/physical-expr/src/datetime_expressions.rs]: it seems that other functions are all using internal_err. If i need to change all of them？

File: datafusion/physical-expr/src/datetime_expressions.rs
Comment (Tangruilin) [reply datafusion/physical-expr/src/datetime_expressions.rs]: Done

File: datafusion/physical-expr/src/datetime_expressions.rs
Comment (Tangruilin) [reply datafusion/physical-expr/src/datetime_expressions.rs]: Done

File: datafusion/physical-expr/src/datetime_expressions.rs
Comment (Tangruilin) [reply datafusion/physical-expr/src/datetime_expressions.rs]: Done

File: datafusion/physical-expr/src/datetime_expressions.rs
Comment (Weijun-H) [reply datafusion/physical-expr/src/datetime_expressions.rs]: Based on https://github.com/apache/arrow-datafusion/pull/8781#discussion_r1444685792, I vote for `exec_err`

Comment (Tangruilin) [reply datafusion/physical-expr/src/datetime_expressions.rs]: So need I change   `internal_err` in  `to_timestamp_invoke`  to `exec_err`?
Or it should be done in another PR?

Comment (Weijun-H) [reply datafusion/physical-expr/src/datetime_expressions.rs]: A follow on pr is better

File: datafusion/expr/src/expr_fn.rs
Comment (Weijun-H) [inline datafusion/expr/src/expr_fn.rs:891]:
```suggestion
    "converts string to date according to the given format"
```

File: datafusion/physical-expr/src/datetime_expressions.rs
Comment (Weijun-H) [inline datafusion/physical-expr/src/datetime_expressions.rs:413]: Maybe we can use `to_timestamp_impl` to avoid code duplication.
```suggestion
        1 => to_timestamp_impl::<TimestampNanosecondType>(args, name),
```

Comment (Tangruilin) [reply datafusion/physical-expr/src/datetime_expressions.rs:413]: We can not do that

for the reason the Date32 need i32 type and time_stamp need i64 type.

to_timestamp_impl did not deal with the converting from i64 to i32

File: datafusion/physical-expr/src/datetime_expressions.rs
Comment (alamb) [inline datafusion/physical-expr/src/datetime_expressions.rs:1396]: It looks like your code takes several input types -- can you please add tests that show calling `to_date` with these types (eg `select to_date(arrow_cast(123, 'Int64'))`)

File: datafusion/sqllogictest/test_files/dates.slt
Comment (alamb) [reply datafusion/sqllogictest/test_files/dates.slt:133]: Can we also please add a test for invalid dates like

```
to_date('21311111')
```

I think the tests here could be copied and adapted to cover `to_date` quite well: 

https://github.com/apache/arrow-datafusion/blob/7641a3228156aab0e48c4bab5a6834b44f722d89/datafusion/sqllogictest/test_files/timestamps.slt#L2073-L2229

**Review by alamb — COMMENTED** [review]
Thank you @Tangruilin  -- this is looking good.

File: datafusion/physical-expr/src/datetime_expressions.rs
Comment (Tangruilin) [reply datafusion/physical-expr/src/datetime_expressions.rs:1396]: Done

**Comment by Weijun-H** [comment]
Hi @Tangruilin , could you add a function description in `scalar_functions.md`?

**Comment by alamb** [comment]
cc @Omega359  do you happen to have some time to give this PR a look?

File: datafusion/sqllogictest/test_files/dates.slt
Comment (Omega359) [inline datafusion/sqllogictest/test_files/dates.slt:175]: This test is repeated - did you mean to change the following tests?

File: datafusion/sqllogictest/test_files/dates.slt
Comment (Omega359) [inline datafusion/sqllogictest/test_files/dates.slt:195]: This test is repeated as well.

File: datafusion/physical-expr/src/datetime_expressions.rs
Comment (Omega359) [inline datafusion/physical-expr/src/datetime_expressions.rs:405]: I still think for one-arg version of this impl we should just use the arrow function to cast from string -> date. Untested but something like this:
`1 => cast_column(&args[0], &DataType::Date32, None),`

File: datafusion/physical-expr/src/datetime_expressions.rs
Comment (Omega359) [inline datafusion/physical-expr/src/datetime_expressions.rs:461]: Can we add some rustdoc with a dataframe based example to this? See [this as an example](https://github.com/apache/arrow-datafusion/blob/main/datafusion/physical-expr/src/regex_expressions.rs#L99)

File: datafusion/physical-expr/src/datetime_expressions.rs
Comment (Omega359) [inline datafusion/physical-expr/src/datetime_expressions.rs:462]: Since there is only a single use of the impl method just inline it here

File: datafusion/physical-expr/src/datetime_expressions.rs
Comment (Omega359) [inline datafusion/physical-expr/src/datetime_expressions.rs:1390]: the validate method should likely be renamed as it won't be used just for timestamps now

**Review by Omega359 — COMMENTED** [review]
Looking good - with a few small updates I think this will be ready.

File: datafusion/common/src/scalar.rs
Comment (alamb) [inline datafusion/common/src/scalar.rs:3243]: Shall we do the same thing for `Date64Type`?

**Review by alamb — COMMENTED** [review]
Thank you @Tangruilin  and @Weijun-H  and @Omega359  for the reviews

@Tangruilin  I think once you have addressed their comments, this PR will be ready to go.

Thanks again

**Comment by alamb** [comment]
Marking this PR as draft as it no longer seems to be waiting on feedback. Please mark it ready for review when it is ready for another look.

**Comment by Omega359** [comment]
@Tangruilin - Do you think you'll have time to complete this PR before v36 is released?

**Comment by Tangruilin** [comment]
> @Tangruilin - Do you think you'll have time to complete this PR before v36 is released?

I will work with it today

**Comment by alamb** [comment]
> > @Tangruilin - Do you think you'll have time to complete this PR before v36 is released?
> 
> I will work with it today

Unfortunately,  I think the 36.0.0 https://lists.apache.org/thread/jhhtz9csocdxx1fzb5ysswjnq79x2fr5 RC has already been cut (so unless we find some problem with RC1 and make a new RC, this issue will have to wait until the next release)

**Comment by Tangruilin** [comment]
> > > @Tangruilin - Do you think you'll have time to complete this PR before v36 is released?
> > 
> > 
> > I will work with it today
> 
> Unfortunately, I think the 36.0.0 https://lists.apache.org/thread/jhhtz9csocdxx1fzb5ysswjnq79x2fr5 RC has already been cut (so unless we find some problem with RC1 and make a new RC, this issue will have to wait until the next release)

I am working with it now.
it is a pity

**Comment by Tangruilin** [comment]
> > > @Tangruilin - Do you think you'll have time to complete this PR before v36 is released?
> > 
> > 
> > I will work with it today
> 
> Unfortunately, I think the 36.0.0 https://lists.apache.org/thread/jhhtz9csocdxx1fzb5ysswjnq79x2fr5 RC has already been cut (so unless we find some problem with RC1 and make a new RC, this issue will have to wait until the next release)

congratulation to 36.0

**Comment by Tangruilin** [comment]
> @Tangruilin - Do you think you'll have time to complete this PR before v36 is released?

You can review it now

**Comment by Tangruilin** [comment]
@Omega359 
You can review this PR now, thanks

**Comment by alamb** [comment]
> @Omega359 You can review this PR now, thanks

@Omega359  once you have reviewed this PR please let me know and I'll give it a final look too and hopefully merge it

Thank you for your work @Tangruilin

File: datafusion-examples/examples/to_date.rs
Comment (Omega359) [inline datafusion-examples/examples/to_date.rs:26]: to_timestamp -> to_date

File: datafusion-examples/examples/to_date.rs
Comment (Omega359) [inline datafusion-examples/examples/to_date.rs:51]: to_timestamp -> to_date, convert 'a to date type

File: datafusion/physical-expr/src/datetime_expressions.rs
Comment (Omega359) [inline datafusion/physical-expr/src/datetime_expressions.rs:437]: Unfortunately we'll have to add an ignore for this because of https://github.com/apache/arrow-datafusion/issues/9277 See https://github.com/apache/arrow-datafusion/pull/9279/files#diff-67ae8808785b2e651767d7ff67dd7c53be04ca098857b52c82ed19927e071cdaR514 for an example. I believe once the functions are move to the new dataframe-functions crate we should be able to re-enable.

File: datafusion/physical-expr/src/datetime_expressions.rs
Comment (Omega359) [inline datafusion/physical-expr/src/datetime_expressions.rs:462]: timestamp => date

File: datafusion/physical-expr/src/datetime_expressions.rs
Comment (Omega359) [inline datafusion/physical-expr/src/datetime_expressions.rs:502]: This should be an exec_err (See https://github.com/apache/arrow-datafusion/pull/9241 for details)

File: datafusion/physical-expr/src/datetime_expressions.rs
Comment (Omega359) [inline datafusion/physical-expr/src/datetime_expressions.rs:1674]: exec_err

**Review by Omega359 — COMMENTED** [review]
Thanks for you work on this @Tangruilin! I left a few comments, mostly minor but 1 issue around the rustdoc where it'll unfortunately have to have 'ignore' added. You have some repeated tests still in dates.slt that should be cleaned up then I believe it's ready to go.

**Comment by Tangruilin** [comment]
@Omega359 The `cargo test --doc`  command is passed on my mac.
But the CI fail. I do not know why

**Comment by Omega359** [comment]
> @Omega359 The `cargo test --doc` command is passed on my mac. But the CI fail. I do not know why

I think it's because you are attempting to include ``` /// # use datafusion::prelude::*; ``` which isn't imported in that module. You can't import it either because of #9277. As mentioned in my review you'll have to ignore that doc test until we move the functions out of this module to the new datafusion/functions area.

File: datafusion-examples/examples/to_date.rs
Comment (alamb) [reply datafusion-examples/examples/to_date.rs:26]:
```suggestion
/// This example demonstrates how to use the to_date series
```

File: datafusion-examples/examples/to_date.rs
Comment (alamb) [reply datafusion-examples/examples/to_date.rs:51]:
```suggestion
    // use to_date function to convert col 'a' to timestamp type using the default parsing
```

**Review by alamb — APPROVED** [review]
I think this PR looks good to me -- thank you @Tangruilin and @Omega359  

There are several cleanup suggestions left outstanding from @Omega359 related to fixing up repeated tests / comments. Is this something you have time to fix @Tangruilin  before we merge the PR? Otherwise we can do it as a follow on PR

**Comment by Tangruilin** [comment]
> I think this PR looks good to me -- thank you @Tangruilin and @Omega359
> 
> There are several cleanup suggestions left outstanding from @Omega359 related to fixing up repeated tests / comments. Is this something you have time to fix @Tangruilin before we merge the PR? Otherwise we can do it as a follow on PR

I think i can fix it in this PR

**Comment by Tangruilin** [comment]
> > I think this PR looks good to me -- thank you @Tangruilin and @Omega359
> > There are several cleanup suggestions left outstanding from @Omega359 related to fixing up repeated tests / comments. Is this something you have time to fix @Tangruilin before we merge the PR? Otherwise we can do it as a follow on PR
> 
> I think i can fix it in this PR

Done

File: datafusion/sqllogictest/test_files/dates.slt
Comment (Omega359) [inline datafusion/sqllogictest/test_files/dates.slt:195]:
```suggestion
```

File: datafusion/sqllogictest/test_files/dates.slt
Comment (Omega359) [inline datafusion/sqllogictest/test_files/dates.slt:211]:
```suggestion
```

**Review by alamb — APPROVED** [review]
Thanks again @Tangruilin  and @Omega359  -- the feature and the review is really appreciated. Nice work

**Pull request merged**
